### PR TITLE
Fixed cube arithmetic for cubes with meshes.

### PR DESCRIPTION
**Pull request by pp-mo**
Has not worked hitherto, because MeshCoords do not behave as AuxCoords in all situations.

I have revised existing code to take account of some cube coords being MeshCoords.

There are still some slight oddities, but I believe these involve behaviours which should be addressed separately -- if at all --,
and not fixed here in a specific way.
Notably :
  1. #4670
  2. a MeshCoord cannot compare equal to a seemingly-matching AuxCoord #4671
  3. a cube cannot be collapsed over a MeshCoord #4672

### CODE REVIEW

**Comment by pp-mo** [comment]
Note: should also fix #4454
We should check that when we have merged this.

**Comment by pp-mo** [comment]
Rebased to pick up #4668

File: lib/iris/tests/unit/analysis/maths/__init__.py [hunk 1]
```diff
@@ -36,8 +38,44 @@ def cube_func(self):
...
+            collapse_dims = set()
+            for co in coords:
+                # Each must produce a single coord, with a single dim
+                (dim,) = cube.coord_dims(co)
+                collapse_dims.add(dim)
+            i_meshdim = cube.mesh_dim()
+            if i_meshdim in collapse_dims:
+                # Make a copy with all mesh coords replaced by their AuxCoord
+                # equivalents.  A simple slicing will do that.
+                slices = [slice(None)] * cube.ndim
```
Comment (pp-mo) [inline lib/iris/tests/unit/analysis/maths/__init__.py:67]: `result = cube[..., 0:]`

File: lib/iris/common/resolve.py [hunk 2]
```diff
@@ -56,10 +58,41 @@
...
+    points: Any
+    bounds: Any
+    dims: Any
+    container: Any
+    mesh: Any = None
+    location: Any = None
+    axis: Any = None
+
+    def create_coord(self, metadata):
+        from iris.experimental.ugrid.mesh import MeshCoord
```
[the remark this answers is not in the document]
Comment (pp-mo) [reply lib/iris/common/resolve.py:74]: It is about circularity yes.
I didn't even check if it was a problem.. I just assumed we would not want this at the top of common/resolve.py

File: docs/src/whatsnew/dev.rst [hunk 3]
```diff
@@ -37,6 +37,9 @@ This document explains the changes made to Iris for this release
 #. `@rcomer`_ implemented lazy aggregation for the
    :obj:`iris.analysis.PERCENTILE` aggregator. (:pull:`3901`)
 
+#. `@pp-mo`_ fixed cube arithmetic operation for cubes with meshes.
+   (:issue:`3107`, :pull:`4651`)
```
[the remark this answers is not in the document]
Comment (pp-mo) [reply docs/src/whatsnew/dev.rst:41]: should be #4554 
(!)

File: lib/iris/common/resolve.py [hunk 4]
```diff
@@ -667,26 +706,86 @@ def _create_prepared_item(
...
+
         prepared_metadata = _PreparedMetadata(
             combined=combined, src=src_metadata, tgt=tgt_metadata
         )
-        bounds = coord.bounds
+
+        if container is None:
+            container = type(coord)
+
+        from iris.experimental.ugrid.mesh import MeshCoord
```
[the remark this answers is not in the document]
Comment (pp-mo) [reply lib/iris/common/resolve.py:748]: I didn't check whether it actually breaks anything, but it seems all wrong to import from iris.experimental.ugrid here.

File: lib/iris/common/resolve.py [hunk 5]
```diff
@@ -56,10 +58,41 @@
...
+    dims: Any
+    container: Any
+    mesh: Any = None
+    location: Any = None
+    axis: Any = None
+
+    def create_coord(self, metadata):
+        from iris.experimental.ugrid.mesh import MeshCoord
+
+        if self.container is not MeshCoord:
```
[the remark this answers is not in the document]
Comment (pp-mo) [reply lib/iris/common/resolve.py:76]: :+1: will do

File: lib/iris/common/resolve.py [hunk 6]
```diff
@@ -1422,30 +1521,68 @@ def _prepare_common_aux_payload(
...
+                        tgt_type = type(tgt_coord)
+                        # Downcast to aux if there are mixed container types.
+                        container = (
+                            src_type if src_type is tgt_type else AuxCoord
+                        )
+                        prepared_item = self._create_prepared_item(
+                            src_coord,
+                            src_metadata=src_metadata,
+                            tgt_metadata=tgt_item.metadata,
+                            dims=tgt_item.dims,
```
[the remark this answers is not in the document]
Comment (pp-mo) [reply lib/iris/common/resolve.py:1575]: :+1: will do

**Review by bjlittle — APPROVED** [review]

**Comment by bjlittle** [comment]
@pp-mo Minor fix required :stuck_out_tongue_winking_eye:

**Pull request merged**